shapes.js

```diff
@@ -0,0 +1,348 @@
+function Shape(x, y, w, h, fill) {
+  this.x = x || 0;
+  this.y = y || 0;
+  this.w = w || 1;
+  this.h = h || 1;
+  this.fill = fill || '#AAAAAA';
+}
+function within_top (shape, mx, my) {
+  return (shape.y + 4 >= my) && (shape.y - 4 <= my)
+}
+
+function within_bottom (shape, mx, my) {
+  return (shape.y + shape.h + 4 >= my) && (shape.y + shape.h - 4 <= my)
+}
+
+function within_right (shape, mx, my) {
+  return (shape.x + shape.w + 4 >= mx) && (shape.x + shape.w - 4 <= mx);
+}
+
+function within_left(shape, mx, my) {
+  return (shape.x + 4 >= mx) && (shape.x - 4 <= mx);
+}
+Shape.prototype.draw = function(ctx) {
+  ctx.fillStyle = this.fill;
+  ctx.fillRect(this.x, this.y, this.w, this.h);
+}
+
+Shape.prototype.contains = function(mx, my) {
+  var minx = Math.min(this.x, this.x + this.w);
+  var maxx = Math.max(this.x, this.x + this.w);
+  var miny = Math.min(this.y, this.y + this.h);
+  var maxy = Math.max(this.y, this.y + this.h);
+  return  (minx <= mx) && (maxx >= mx) &&
+          (miny <= my) && (maxy >= my);
+}
+
+Shape.prototype.border = function (mx, my) {
+  //Checks if within border
+  if (within_top(this, mx, my) && within_left(this, mx, my)) {
+    return Shape.prototype.borderenum.TOPLEFT;
+  } else if (within_top(this, mx , my) && within_right (this, mx, my)) {
+    return Shape.prototype.borderenum.TOPRIGHT;
+  } else if (within_bottom(this, mx, my) && within_right(this, mx, my)) {
+    return Shape.prototype.borderenum.BOTTOMRIGHT;
+  } else if (within_bottom(this, mx, my) && within_left(this, mx, my)) {
+    return Shape.prototype.borderenum.BOTTOMLEFT;
+  } else if (within_top(this, mx, my)) {
+    return Shape.prototype.borderenum.TOP;
+  } else if (within_left(this, mx, my)) {
+    return Shape.prototype.borderenum.LEFT;
+  } else if (within_right(this, mx, my)) {
+    return Shape.prototype.borderenum.RIGHT;
+  } else if (within_bottom(this, mx, my)) {
+    return Shape.prototype.borderenum.BOTTOM;
+  } else {
+    return false; 
+  }
+}
+
+Shape.prototype.borderenum = {
```

Comment:
You can get away with using a string instead of an enum. Even if enum is faster, we want good code over slightly higher performance unless we know that it's causing performance problems.